test/unit/warp/src/Shfl.cpp

```diff
@@ -0,0 +1,130 @@
+/* Copyright 2021 David M. Rogers
+ *
+ * This file is part of Alpaka.
+ *
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+
+#include <alpaka/test/KernelExecutionFixture.hpp>
+#include <alpaka/test/acc/TestAccs.hpp>
+#include <alpaka/test/queue/Queue.hpp>
+#include <alpaka/warp/Traits.hpp>
+
+#include <catch2/catch.hpp>
+
+#include <cstdint>
+
+#define CAST(x) (*((std::int32_t*) (&(x))))
+
+//#############################################################################
+class ShflSingleThreadWarpTestKernel
+{
+public:
+    //-------------------------------------------------------------------------
+    ALPAKA_NO_HOST_ACC_WARNING
+    template<typename TAcc>
+    ALPAKA_FN_ACC auto operator()(TAcc const& acc, bool* success) const -> void
+    {
+        std::int32_t const warpExtent = alpaka::warp::getSize(acc);
+        ALPAKA_CHECK(*success, warpExtent == 1);
+
+        ALPAKA_CHECK(*success, alpaka::warp::shfl(acc, 12, 0) == 12);
+        ALPAKA_CHECK(*success, alpaka::warp::shfl(acc, 42, -1) == 42);
+        float ans = alpaka::warp::shfl(acc, 3.3f, 0);
+        float expect = 3.3;
+        ALPAKA_CHECK(*success, CAST(ans) == CAST(expect));
```

Comment:
While that may work on most compilers, it is certainly undefined behavior because we are reading a float value as an int32. At that point a `memcmp` is probably the more correct solution.

However, I think I prefer disabling the warning here. I think we know what we are doing here and we expect the exact same float to be shuffled around, so the comparison should not need an epsilon. 

```c++
#pragma GCC diagnostic push
#pragma GCC diagnostic ignored "-Wfloat-equal"
ALPAKA_CHECK(*success, ans == expect);
#pragma GCC diagnostic pop
```

I don't know if we need a similar escape hatch for the other compilers.

Reply:
Compiler specific pragmas should be guarded: see https://github.com/alpaka-group/alpaka/blob/a78a70adb27c5285fd8752f96c11db1487a20906/include/alpaka/atomic/Op.hpp#L31-L34